### ISSUE TITLE
"MS Word opening of big .docx file" slower on mirror than on normal file system (C:\)

### ISSUE THREAD

**Issue by cdarau** [dokan-dev/dokany#521]
# Environment
 * Windows version: Window 10 Enterprise
 * Processor architecture: x64
 * Dokany version: 1.0.3.1000
 * Library type (Dokany/FUSE): Dokany
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x] I can always reproduce the issue with the provided description below.
  - [x] I have updated Dokany to the [latest version](https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
  - [x] I tested one of the last [snapshot](https://github.com/dokan-dev/dokany/wiki/Build#user-snapshot) from appveyor CI

# Description
When I open a big .docx file with MS Word, I see a big difference in time between the case when the file is served through the mirror and the case when it is served through the normal file system (a desktop folder for example).

I opened a 35.9 MB .docx file with MS Word, first served by mirror and then from a desktop folder, and I got the following results:
- ~98 seconds between first and last ReadFile operation on the file from the mirror
- ~45 seconds between first and last ReadFile operation on the file from the normal file system

The special part is that MS Word does some >300k small reads (100 bytes each to be exact) and I think that this behavior triggers the observable latency in the mirror case. It appears that somewhere between entering the dokany kernel and reaching the callbacks there is a problem. Not in the mirror callbacks themselves.

# Logs
There are some problems with uploading the logs. I've experienced these problems before. Although it says that it allows zip files, it doesn't work. So I added the log files to my drive and shared the link here.

<deleted>

**Comment by Liryna** [comment]
Hi @cdarau ,

Can it be the same behavior as https://github.com/dokan-dev/dokany/issues/515 ? (little read size compared to normal fs)

I cannot see the log files on the drive, it ask me to login with a simplision.com mail. You can upload the log on github by renaming it to `.txt`

**Comment by cdarau** [comment]
Hello @Liryna 

I renamed the .zip file to a .txt file and it worked.

I checked the issue (#515) but it isn't the same thing as here. There, when opening the WinRAR file there were far more reads on the mirror scenario than on the windows one. Here, the sequences are exactly the same. Only the speed differs (mirror performing 2x slower when serving those reads). So this issue is about some problem in the dokan layers that make them serve so slow, comparing to the windows scenario. When using even bigger files, the difference is even bigger, overpassing 2x ratio. Maybe the short length (only 100 bytes) of those hundreds of thousands reads is the cause, I don't know.

[DesktopZIP.txt](https://github.com/dokan-dev/dokany/files/1043275/DesktopZIP.txt)

**Comment by Liryna** [comment]
Hello @cdarau ,

If read/write speed can depend on the number of thread used in dokan library. How much are you using now ?
If you are willing to make more test about speed. Take a look at `Corillian-asyncio` there is no longer this thread logic.
https://github.com/dokan-dev/dokany/tree/Corillian-asyncio
For this would need to build the snapshot build:
https://ci.appveyor.com/project/Maxhy/dokany/build/1.0.3-189/job/khguw3b41v1pemfd/artifacts

**Comment by cdarau** [comment]
Hello @Liryna 

I'm using the maximum no, 14.

This branch you're referring to is interesting, but unfortunately I need a signed driver :( Is this anywhere in sight for this second branch?

Thank you very much.

**Comment by Liryna** [comment]
I will try to make a beta of 2.0.0 but this will only be in a 1-2 week 😢 
Ping me in case I forget.

**Comment by cdarau** [comment]
Hello @Liryna 

Just adding this comment in order to ping you as you said :-)

Also, I wanted to ask you something important: does the upcoming new dokany has the fuse wrapper? I see that you're updating the major no so I except some big differences. Beside this thread logic change that will characterize the opcoming 2.0.0, what significant other changes to except? (I refer to those that would break compatibility)

Thank you very much!

**Comment by SergiuPol** [comment]
Hopefully 2.0 will solve the performance problem met in scenarios with large number of IOs like opening large DOCX and PDF. Performance is a major concern and great challenge for those who write serious code.

**Comment by Liryna** [comment]
Hi @cdarau ,

Thanks for the ping ! Will have to delay the beta 😢 Currently still under a lot of work.

Yes, the upcoming new dokany has the fuse wrapper but has really not been tested. 2.0.0 is really at an early test stage. It crash, it needs more testing but the speed is here !
See https://github.com/dokan-dev/dokany/pull/307#issuecomment-300071950 and other comments for more information or if you have other questions about the 2.0.0.

@SergiuPol I agree that speed is important ! Dokan is already able to support heavy read/write (some project manage heavy video files without problems).

The behavior of this issue is not related directly to Dokan speed but probably more by an unexpected behavior from Dokan to the Kernel that make slow down the read/write (reduce buffer size and therefore grow unnecessarily the number of IO). If the behavior that creates the issue is the speed, then yes 2.0.0 will fix it. Otherwise we will need to look more into it !

@Corillian who rewrote dokan library for 2.0.0 is a serious guy who writes serious code 👍 
Dokan ❤️  serious guys, so do not hesitate !

**Comment by Corillian** [comment]
Thank you for the kind words @Liryna, I appreciate it =). @SergiuPol while the work I did on async IO for dokan 2.0 goes a long way toward improving performance there are still a number of areas where significant enhancements could be made. Unfortunately these all center around non-trivial kernel changes that would also likely require an additional set of significant changes to the user mode driver. For starters the kernel mode driver serializes all IO requests to a single queue for handling by the user mode driver. This not only creates a single point of contention across _all_ IO requests but it also requires a large number of buffer copies to take place during reads and writes. I would still need to perform additional research but I believe this could be heavily mitigated by using APC's between the kernel driver and the user driver. A less performant but less disruptive solution would be supporting a single queue per file-handle with a thread pool for servicing requests - similar to what my dokan 2.0 user mode code is doing.

Unfortunately without the participation of more people with kernel level development experience these changes may be a long time in coming. A technical roadmap would likely need to be created with proper debate around the best path to take to address dokan's most significant shortcomings.

**Comment by SergiuPol** [comment]
@Corillian I would like to help but I have sysadmin background (thus Dokany power user) not a developer :)
Appreciating the depth of your work, I don't dare making suggestions about writing code.

When sysadmins meet such problems, as they cannot deep dive into the code to produce improvements there, typically they look for alternatives. As far as I know, this is rarely a good approach for developers. But who knows, maybe you find some ideas here:
https://sourceforge.net/p/ext2fsd/code/ci/master/tree/
https://github.com/Microsoft/GVFS

As far as I know, both the file systems above use the callback mechanism.

**Comment by Rondom** [comment]
Coming back to the issue: Don't you think something else goes wrong here? Why does Word make so small requests? Dokan can do nothing against a userspace-application making inefficient read-requests?

**Comment by SergiuPol** [comment]
@Rondom sure it does! No doubt msf word and foxit pdf reader (probably many others) are io pigs :)

I see not a single reason for reading an entire document in many srwuential small reads instead of a few large reads, given the fact that viewers and editors typically fetch the entire file in memory. Thus I do not defend userspace applications writen amateuristic :)

Oh yes dokan can do something - forward all those requests asap to the back-end fs.

Let's put this in another light: dokan - that is not a full-blown fs (that treats the file table, chunks, caching, journaling and fragmentation problems) but is more like a fs proxy - doubles the docx/pdf opening time. Isn't this perf penalty too much for its limited purpose ? :)

Don't get me wrong. I do appreciate the work behind dokan! For many other use cases, it performs good. I am just ofering  my non-dev contribution to make dokan even better by providing feedback and suggestions from another view point.

_Sent from my HTC_M10h using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Rondom** [comment]
I did not express myself propely. I am not saying the app is at fault. But somehow Dokan makes the app do un-optimal reads. I mean it works fine with NTFS, why does an application change its behaviour? Ideally we can find an open source App with the same behaviour to simplify the analysis...

**Comment by SergiuPol** [comment]
As far as I know, msf word io pattern is the same in the case of dokan and ntfs, just that dokan + mirror + ntfs is 2 times slower than ntfs directly.

Thus I would not spend time investigating the userspace app but the time spent in dokan.

But if you still want to be very, very sure that the root cause is in dokan, maybe @cdarau can write a few-lines app that reads the entire input file in small chunks and prints the total read time. Isolating the issue in this manner will give us maximum clarity.

What do you think?

_Sent from my HTC_M10h using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Rondom** [comment]
Oh sorry! I mixed up issues! As @cdarau has said above the behaviour is the same for Word on NTFS and Dokan (i.e. same read-patterns, only the performance is worse on Dokan).
Ignore the two comments above. I wrote them thinking this was issue #515 
Now I can also see the reason for your confusion! Sorry! I should sleep!

**Comment by cdarau** [comment]
Hello to all.

I just wanted to add that even when serving from RAM, the time-wise poor performance is noticed. So the problem doesn't reside in the mirror-callbacks, but somewhere before them in the dokany layers.

**Comment by Corillian** [comment]
@cdarau as explained above:

> @SergiuPol while the work I did on async IO for dokan 2.0 goes a long way toward improving performance there are still a number of areas where significant enhancements could be made. Unfortunately these all center around non-trivial kernel changes that would also likely require an additional set of significant changes to the user mode driver. For starters the kernel mode driver serializes all IO requests to a single queue for handling by the user mode driver. This not only creates a single point of contention across all IO requests but it also requires a large number of buffer copies to take place during reads and writes. I would still need to perform additional research but I believe this could be heavily mitigated by using APC's between the kernel driver and the user driver. A less performant but less disruptive solution would be supporting a single queue per file-handle with a thread pool for servicing requests - similar to what my dokan 2.0 user mode code is doing.

**Comment by SergiuPol** [comment]
@Corillian The above issue refers to 1.0.3. After the signed 2.0 build will be ready, we will check if the problem is still there and let you know.

**Comment by Liryna** [comment]
Sorry for delay ! Here is the first beta: https://github.com/dokan-dev/dokany/releases/tag/v2.0.0-BETA1

**Comment by SergiuPol** [comment]
Great news! :)

Does this build solve the perf problem in the scenario with many IO operations?

Does it also send to the user-mode the lock type?

10x!
Sergiu.

**Comment by Liryna** [comment]
@SergiuPol This release is exactly what @Corillian said previously.
It really improves many IO operations in the library side. There is no longer limitations at thread numbers. That really enhances a lot the speed.
There is still the kernel <-> userland limitation that @Corillian explained very well.
I do not think we will be able to make this change in 2.0.0 or if we get a lot of support by the community to speed up the 2.0.0 

https://github.com/dokan-dev/dokany/pull/307#issuecomment-300071950

**Comment by cdarau** [comment]
Hello again,

I just tested the new release, namely 2.0.0.BETA1, from corillian-asyncio branch. Unfortunately, there is the same problem when opening a big .docx file and Word decides to launch on those 100s of thousands small reads (100 bytes).

![image](https://user-images.githubusercontent.com/9717639/32499981-b9832cb2-c3dc-11e7-89d5-3b35d54c5cd3.png)

So what I did:
- I had 3 files: the 1st one ~55MB, the 2nd one ~36MB, the 3rd one ~55MB
- I opened each one with both the mirror and the ntfs fs
- I repeated the above 6 openings 3 times
- the results show that when opening the 36MB .docx file (the one that triggers those small reads) with the mirror I get a time >10x slower than that obtained with the ntfs fs

@Corillian what can possibly cause this behaviour?

Thank you all
